Add mapgen settings to create world dialog

srifqi opened the pull request:
Any changes in mainmenu, change the minetest.conf.
Please run `updatepo.sh` after merge.

~~Require #2565 to run.~~ (fixed)
- [x] Better name for "Light" (solved, thanks to @Rui914)
- [x] Found a bug, player cannot re-check checkbox after save (solved, ~~require #2565~~)
- [x] Adapt to new mapgens and settings

![minetest-201603051219 mainmenu mapgen v6 flags dialog](https://cloud.githubusercontent.com/assets/4017302/13545943/ef701df2-e2cf-11e5-9913-4f0391a69d5f.PNG)

srifqi committed:
Add mapgen settings to create world dialog

Any changes in mainmenu change the minetest.conf.
Supports new mapgens.

srifqi commented:
The old one (#2480) has deleted :(
So, I create another pr

Zeno- commented:
This is starting to look good.  The only problem is it doesn't seem to work for me :)

C1ffisme commented:
Screenshots?

srifqi commented:
![minetest-201503270609 mainmenu create world dialog flags](https://cloud.githubusercontent.com/assets/4017302/6859135/33acfc70-d448-11e4-864f-8446e60173fe.PNG)
^ The Create World dialog

![minetest-201503270611 mainmenu mapgen flags dialog](https://cloud.githubusercontent.com/assets/4017302/6859143/4192b8c0-d448-11e4-9778-05aa2e6404d9.PNG)
^ The mapgen dialog

C1ffisme commented:
:+1:

srifqi commented:
Updated,
Add `mg_spflags` param a.k.a. Generated structure
Require #2565 to merge

srifqi commented:
Updated,
Just know about `core.is_yes`. So, I removed #2565 (`toboolean`). Also, the code has changed.
Thanks, @est31!

Rui914 commented:
How is "Sunlight"?

srifqi commented:
@Rui914 Nice! Will update ASAP.

srifqi commented:
Updated, thanks to @Rui914.

Rui914 commented:
When the other person is the OK, it will be squash.

An error has occurred in Travis, but it is okay to not have to worry about.

srifqi commented:
Quick update:
- Always load default mapgen flags at start.
- Change "Flags" to "Settings"

rubenwardy commented:
+1

nerzhul commented:
jenkins rebuild please

nerzhul commented:
jenkins build please

srifqi commented:
Okay, squashed.
Ready to merge.

nerzhul commented:
We are waiting the builds :)

Rui914 commented:
I can't wait :)
The next release is a good release!

Rui914 commented:
I can't wait :)
The next release is a good release!

srifqi commented:
Um... @Rui914 You commented twice :D

Also, why this pr does not have any labels?

nerzhul commented:
Because nobody set it ?

srifqi commented:
Okay, @nerzhul Thanks!

Also, all build has passed, successful checks.

fz72 commented:
Why did you create a save button but no cancel button. Maybe you should add a save button and a cancel button. Only set the settings on save and do nothing on cancel.

fz72 commented on builtin/mainmenu/dlg_create_world.lua:
```diff
@@ -15,6 +15,12 @@
+core.setting_set("mg_flags","trees, caves")
+core.setting_set("mgv5_spflags","blobs")
+core.setting_set("mgv6_spflags","biomeblend, jungles, mudflow")
+core.setting_set("mgv7_spflags","mountains, ridges")
```
Why not loading the default settings from minetest.conf?

srifqi replied:
We have to load the default, so any player will create world without worrying about the parameter.

fz72 replied:
We could load the default from minetest.conf. At beginning these are the defaults from minetest.conf.example. When you change the flags in the dialog you could store them in minetest.conf. Then the players have always their own default settings.

srifqi replied:
Nope. But you can make another pr when this get merged.

fz72 commented on builtin/mainmenu/dlg_create_world.lua:
```diff
@@ -127,8 +133,23 @@ local function create_world_buttonhandler(this, fields)
+	end
 
-	return false
+	return ddhandled
```
why you need ddhanled?

srifqi replied:
See at line 146-150. There is a dropdown handler or, can I remove it?

fz72 replied:
I see it. But you return true in the dropdown. And only the dropdown is not handled you need to return false. I think you don't need ddhandled.

srifqi replied:
I leave it there for another things, if exist.

srifqi commented:
@fz72
> Why did you create a save button but no cancel button. Maybe you should add a save button and a cancel button. Only set the settings on save and do nothing on cancel. Maybe the dialog should look like the others with the save and cancel button at the bottom.

I forgot that, because it always save anything every change.
Thanks,for reminding me!

srifqi commented:
Updated, thanks @fz72!

Squashed. Ready to merge.

Megaf commented:
Well done @srifqi 
@nerzhul @Zeno- 
I'd like seeing this merged if you agree with the coding style and don't find any obvious bug.
:+1:

Rui914 commented:
:+1: !

est31 commented on builtin/mainmenu/dlg_mg_flags.lua:
```diff
@@ -0,0 +1,258 @@
+	-- generated structure
+	-- trees, caves, flat, dungeons, light.
+	local mg_flags = ""
+	if(flag_list["trees"] == true) then
```
This has alot of copypasta. Why not do something like this instead? 
```
local first = true
for flag, val in pairs(flag_list, flag_str) do
	flagstr = flagstr .. (first and "" or ", ") .. ((val == true) and flag)
	first = false
end
```
Put that code into a helper function, and separate the flag_list into normal flags, mgv6 flags and mgv7 flags. `mg_flags_buttonhandler` can be similarly shortened.

srifqi replied:
> This has alot of copypasta. Why not do something like this instead?

Thanks! Will update ASAP. But, should the mgv5_spflags also use it?

> `mg_flags_buttonhandler` can be similarly shortened.

How?

est31 replied:
Something like this:
```function helper_func(flag_list)
	for flag, val in pairs(flag_list) do
		local cbflag = "cb_" .. flag
		if fields[cbflag] then
			flag_list_g[flag] = core.is_yes(fields[[cbflag])
			update_mg_flags()
			return true
		end
	end
	return false
end

if helper_func() then
	return true
end```

srifqi commented:
Updated. Thanks, @est31!

Squashed. Ready to merge.

nerzhul commented:
It's okay for me now ! Great job

est31 commented:
@srifqi Have you seen my comment on how to shorten `mg_flags_buttonhandler`?

kilbith commented:
Right now, it miss the snow biome in your options : https://github.com/minetest/minetest/commit/75cbd80e5b78af52eae2ec9960d408e64d33adca

srifqi commented:
@est31 Thanks for answering my question. Will update ASAP.
@kilbith Since when there are snow biomes in Minetest? Is it `Mapgen flags`, or `Generated structure`?

everamzah commented:
'mgv6_spflags = snowbiomes' and this turns on jungles.

Calinou commented:
> 'mgv6_spflags = snowbiomes' and this turns on jungles.

Call the checkbox "Additional biomes" maybe.

srifqi commented:
@everamzah
> 'mgv6_spflags = snowbiomes' and this turns on jungles.

Turn on jungles? If it's also add jungle, should I still add `jungles` and `nojungles`?

@Calinou Nice idea! (only if ALSO turn on jungles)

kilbith commented:
@srifqi Since yesterday.

paramat commented:
The 'snowbiomes' option should indeed be under 'v6 mapgen flags' alongside 'biomeblend' 'jungles' and 'mudflow', because it is also a v6-specific flag.
When snowbiomes are enabled there is no choice about jungles, jungles are force-enabled.
Nice PR, please wait for kwolekr (hmmmm) to approve it as he's the mapgen manager and the expert.

kwolekr commented:
The light flag should not be exposed to the end user.  This is intended for mods only - I cannot conceive of any circumstance where a player would actually want to turn off lighting.
The Mapgen v5 'blobs' flag is deprecated.
If you're going to add mapgen flags, then why not other settings such as water level?
Also, can we get an updated screenshot?

paramat commented:
> core.setting_set("mg_flags","trees, caves")

The default mg_flags are stored in mapgen.h and are actually 'trees, caves, light'.

> We could load the default from minetest.conf. At beginning these are the defaults from minetest.conf.example.

Minetest.conf doesn't store the default flags, the default 'mg_flags' are stored in 'mapgen.h', the default 'mgv5/6/7_spflags' (mapgen specific flags) are stored in 'mapgen_v5/6/7.cpp'.
I think the flag selections made in the create world menu should only be for the world about to be created, they should not be written to minetest.conf.

paramat commented:
> core.setting_set("mg_flags","trees, caves")

The default 'mg_flags' are stored in the 'src/mapgen.h' file and are actually 'trees, caves. light'.

> We have to load the default (from: minetest.conf.example), so any player will create world without worrying about the parameter.

^ Yes. But how does this PR work with minetest.conf? The default flags should be read from 'core.setting_set' and then should be modified by any flags enabled or disabled in minetest.conf.
Does the 'OK' button save the chosen flags to minetest.conf?

paramat commented:
> We have to load the default (from: minetest.conf.example), so any player will create world without worrying about the parameter.

How does this PR work with 'minetest.conf' and the 'minetest.conf' of the subgame?
The flags that appear as default in your menu should first be read from 'core.setting_set', then modified by any flags enabled or disabled in the 'minetest.conf' of the subgame, then finally modified by any flags enabled or disabled in the engine's 'minetest.conf'.

Does the 'OK' button save the chosen flags to 'minetest.conf'? I think it should not, as 'minetest.conf' stores the player's general preferences, the player will often want to make each specific world different from those general prefernces.

srifqi commented:
@kwolekr
> The light flag should not be exposed to the end user. This is intended for mods only - I cannot conceive of any circumstance where a player would actually want to turn off lighting.

Let just put it there, just in case.

> The Mapgen v5 'blobs' flag is deprecated.

Oh, okay. I'll remove it.

> If you're going to add mapgen flags, then why not other settings such as water level?

Good idea! Can you give list of setting that needs to add here?

> The code itself looks very repetitive. Could you change this to use a loop and a field name lookup table instead?

I'm trying to do that.

> Can we get an updated screenshot?

Sorry, not now. Maybe later.

@paramat
> >We have to load the default (from: minetest.conf.example), so any player will create world without worrying about the parameter.

> ^ Yes. But how does this PR work with minetest.conf? The default flags should be read from 'core.setting_set' and then should be modified by any flags enabled or disabled in minetest.conf.
Does the 'OK' button save the chosen flags to minetest.conf?

For now, it ALWAYS save any changes into `minetest.conf`. Always load default. No cares about settings from `minetest.conf`. I'm trying to add support for reading the `minetest.conf`

srifqi commented:
Okay, updated.
Squashed.

paramat commented:
> For now, it ALWAYS save any changes into minetest.conf

That's okay by me.

est31 commented on builtin/mainmenu/dlg_create_world.lua:
```diff
@@ -127,8 +127,23 @@ local function create_world_buttonhandler(this, fields)
 		this:delete()
 		return true
 	end
+	
```
Please remove the trailing whitespaces

est31 commented on builtin/mainmenu/dlg_mg_flags.lua:
```diff
@@ -0,0 +1,211 @@
+		return true
+	end
+	
+	function helper_func(flag_list)
```
I've found an even more compact way of writing this. Instead of a function you do an outer for loop over a previously declared `local flag_lists = {flag_list_g, flag_list_6, flag_list_7}`

srifqi replied:
Yup! Nice idea!
I always like golfing in programming.

est31 commented on builtin/mainmenu/dlg_mg_flags.lua:
```diff
@@ -0,0 +1,211 @@
+	-- this is it!
+	local formspec =
+		"size[8,4,true]" ..
+		"button[4,3.65;3,0.5;mg_ok;" .. fgettext("Ok") .. "]" ..
```
I think "Close" is a better name for this, as the changes take effect already after the person has made the change.

srifqi replied:
Okay, agreed.

srifqi commented:
Okay, updated.
Many changes from @est31.

Squashed.

est31 commented on builtin/mainmenu/dlg_mg_flags.lua:
```diff
@@ -0,0 +1,221 @@
+		update_mg_flags()
+	end
+	mg_first_time = false
+	
```
There are still trailing whitespaces

est31 commented on builtin/mainmenu/dlg_mg_flags.lua:
```diff
@@ -0,0 +1,221 @@
+		if helper_func(flag_list[i]) then
+			return true
+		end
+	end
```
I rather had something like this in mind:
```
	local flag_list = {flag_list_g, flag_list_6, flag_list_7}
	for idx, flag_list in pairs(flag_lists_list)
		for flag, val in pairs(flag_list) do
			local cbflag = "cb_" .. flag
			if fields[cbflag] then
				flag_list[flag] = core.is_yes(fields[cbflag])
				update_mg_flags()
				return true
			end
		end
	end
```

srifqi commented:
Okay, updated.
More trick by @est31

Squashed.

srifqi commented:
Just a note:

Please run `updatepo.sh` after merging.

Gael-de-Sailly commented:
:+1:

Dragonop commented:
:+1:

paramat commented:
Is it still needed though now we have a new settings menu?

RHRhino commented:
@paramat The new settings menu with all its submenus is a pain if you have to change a lot of settings e.g. if you want to change the FPS or shaders you have to go though 5 submenus :angry: 
Mapgen also has at least 2-3. I think new player definitely need a simple way to change their settings.

Gael-de-Sailly commented:
That has been said many times on a topic on the forum, [since this post](https://forum.minetest.net/viewtopic.php?p=195497#p195497).
But we still need a separate per-world conf dialog when creating worlds. That's way more easy to use than going to the settings before creating a new world.

paramat commented:
Yes i somewhat agree, even PilzAdam admits the new settings menu needs a simpler version alongside.

paramat commented:
The global 'flat' and 'trees' flags have now been moved into mgv6 spflags and there's a new global 'decorations' flag.

srifqi commented:
Uh, can anybody make a list so that I can understand what's new? Because I was doing other things - not related to Minetest.

paramat commented:
Apart from what i mentioned above ..
We have a new settings menu that uses data from https://github.com/minetest/minetest/blob/master/builtin/settingtypes.txt you might want to load in flag data from there instead.
We have 3 new mapgens.

Also i agree with kwolekr/hmmmm that the 'sunlight' flag should not be present here, players almost never need to set that, that's for use by lua mapgen mods.

kwolekr > The light flag should not be exposed to the end user. This is intended for mods only - I cannot conceive of any circumstance where a player would actually want to turn off lighting.

srifqi commented:
Finally, it includes new mapgens (and flags).
Hasn't rebased.

> Call the checkbox "Additional biomes" maybe.

Done. :smile:

Here are some sceenshots:

![minetest-201603051219 mainmenu mapgen v6 flags dialog](https://cloud.githubusercontent.com/assets/4017302/13545943/ef701df2-e2cf-11e5-9913-4f0391a69d5f.PNG)
^ Mapgen [v6] Flags Dialog

![minetest-201603051220 mainmenu mapgen v6 flags dialog no additional biomes](https://cloud.githubusercontent.com/assets/4017302/13545950/281db1a0-e2d0-11e5-9948-1698e5c306ec.PNG)
^ Mapgen [v6] Flags Dialog (no Additional Biomes)

![minetest-201603051220 mainmenu mapgen v7 flags dialog](https://cloud.githubusercontent.com/assets/4017302/13545951/3495de26-e2d0-11e5-8567-ea7329785c03.PNG)
^ Mapgen [v7] Flags Dialog

![minetest-201603051221 mainmenu mapgen flat flags dialog](https://cloud.githubusercontent.com/assets/4017302/13545955/4008cbba-e2d0-11e5-8666-017078feba59.PNG)
^ Mapgen [flat] Flags Dialog

![minetest-201603051222 mainmenu mapgen valleys flags dialog](https://cloud.githubusercontent.com/assets/4017302/13545960/4a1ba92e-e2d0-11e5-85e5-62f149f13ff1.PNG)
^ Mapgen [valleys] Flags Dialog

paramat commented:
I'll try to test soon.

paramat commented:
Please squash and rebase, then i'll test.

Gael-de-Sailly commented:
:+1:

srifqi commented:
Squashed just now. a7452bd1607d5eb0574ea460f700d7bac548bc3f

paramat commented:
Is it rebased? it failed to apply in testing @srifqi

srifqi commented:
> Is it rebased? it failed to apply in testing

Sorry, @paramat. I only squash it. I'm trying to rebase now.

paramat commented:
Saw your message, i've been busy but today testing this is top priority.

paramat commented:
There are 12 whitespace errors.

When i access the mapgen world settings page, all options are enabled by default, they don't seem to read the default flag settings of each mapgen. In many mapgens some of the sp_flags are disabled by default.

If i start a world without accessing the world settings page, all sp_flags are enabled, so mgv6 becomes flat, which is not the default.
When i then exit the game, all mapgen flags for all mapgens are written to .conf as being enabled.

srifqi commented:
Thanks for your review!

First, can you please spot where are those whitespace error? I've read the [Lua code style guidelines](http://dev.minetest.net/Lua_code_style_guidelines), but seems like I'm missing something.

Second,it is a little bit makes me confused. I'll try again and learn more to make sure that I'm understand that it really works.

paramat commented:
In mgv6 'flat' is not enabled by default.
In mgflat the double space in the line means that the default is "" == empty.
Best to look in conf.example for the defaults it's clearer:

mg_flags = caves,dungeons,light,decorations
mgv6_spflags = jungles,biomeblend,mudflow,snowbiomes,trees
mgv7_spflags = mountains,ridges
mgflat_spflags = "" (empty string)
mg_valleys_spflags = altitude_chill,humid_rivers

srifqi commented:
> Also please could you rename 'additional biomes ' to 'snowbiomes' to match the flag?

I think no. Because actually, it auto-enables jungle biome.


![ground level on flat mapgen](https://cloud.githubusercontent.com/assets/4017302/13980745/af4f3b6c-f112-11e5-850b-05de4bb94c42.png)
I also add `Ground level` into flat mapgen.

Rebased and squashed.

paramat commented:
> I think no. Because actually, it auto-enables jungle biome.

Okay.

paramat commented:
Please can you remove 'ground height' from mgflat? I feel these menus should only be for mapgen flags, otherwise there will be too many other settings included. Ground height is a more advanced setting so can stay in the advanced setting menu.

> There is "water/lava spawn in cave" in valleys mapgen configuration. Should I add them too?

No.

paramat commented:
> Then when i then exit the game, ... flags for some or all mapgens are written to .conf as being enabled, which mapgens have their flags written seems random and changes each time.

I'm still concerned about this behaviour, it is different to current behaviour.

srifqi commented:
> Please can you remove 'ground height' from mgflat?

Done.

> > Then when i then exit the game, ... flags for some or all mapgens are written to .conf as being enabled, which mapgens have their flags written seems random and changes each time.

> I'm still concerned about this behaviour, it is different to current behaviour.

So, this is what's going on to `minetest.conf` (was clean) after running Minetest.
```
mg_flags = decorations, caves, light, dungeons
mg_valleys_spflags = humid_rivers, altitude_chill
mgflat_spflags = nolakes, nohills
mgv6_spflags = mudflow, biomeblend, trees, jungles, snowbiomes, noflat
mgv7_spflags = ridges, mountains
water_level = 1
```
Because I use boolean to save temporary data, only yes or no (`lakes` or `nolakes`) is save for each flag in `minetest.conf`.

Rebased and squashed.

paramat commented:
What concerns me is that currently flags are not written to .conf if they are not altered from default, i am no expert on menu settings but it seems better to keep the current behaviour.
@PilzAdam any comments on this?

paramat commented:
This might be affected by upcoming big changes to mapgen settings by hmmmm / kwolekr.

paramat commented:
Looks like the big changes to mapgen settings may not happen, hmmmm is unsure.

0-afflatus commented:
There will always be upcoming changes. It's a good idea and looks ready to be put through its paces by real players. The only outstanding issue seems to be the menu settings behaviour.

paramat commented:
Needs updating for mgv7 floatlands, caverns and mgv5 caverns, but there may be other issues to discuss first.

srifqi commented:
Has the API documentation updated?

SmallJoker commented on builtin/mainmenu/dlg_mg_flags.lua:
```diff
@@ -0,0 +1,243 @@
+}
+
+-- valleys mapgen
+local flag_list_v = {
```
`flag_list_valleys`, same logic for the other table definitions, please.

SmallJoker commented on builtin/mainmenu/dlg_mg_flags.lua:
```diff
@@ -0,0 +1,243 @@
+local flag_list_7 = {
+	mountains	= true,
+	ridges		= true
+}
```
What paramat meant: `floatlands` and `caverns` are missing here
https://github.com/minetest/minetest/blob/master/minetest.conf.example#L1339

SmallJoker commented on builtin/mainmenu/dlg_mg_flags.lua:
```diff
@@ -0,0 +1,243 @@
+
+-- Load default settings if not exist
+if core.setting_get("water_level") == nil then
+	core.setting_set("water_level", 1)
```
No need to set the defaults or checking for the existence. Please correct me when I see this one the wrong way but the default value is loaded if there's one defined in `defaultsettings.cpp`.

SmallJoker commented on builtin/mainmenu/dlg_mg_flags.lua:
```diff
@@ -0,0 +1,243 @@
+		else
+			formspec = formspec ..
+			"checkbox[0.25,2.0;cb_jungles;" .. fgettext("Jungles") .. ";" ..
+				dump(flag_list_6["jungles"]) .. "]"
```
The output of `dump` is not garantueed to have a fixed format. Please use `tostring` or a custom function, if needed here.

SmallJoker commented on builtin/mainmenu/dlg_mg_flags.lua:
```diff
@@ -0,0 +1,243 @@
+
+function create_mg_flags_dlg()
+	local retval = dialog_create("sp_mg_flags",
+					mg_flags_formspec,
```
Return directly.

SmallJoker reviewed (dismissed):
See comments.

paramat commented:
"What concerns me is that currently flags are not written to .conf if they are not altered from default, i am no expert on menu settings but it seems better to keep the current behaviour."
As in, current behaviour is that flags are not written to .conf if not altered from default, this is how all settings work in .conf so should be preserved.
:-1:

srifqi commented:
So, should this write to ```minetest.conf``` even if it hasn't been changed, or not?

paramat commented:
Not.

paramat commented:
IRC:
19:52 paramat     the behaviour is wrong, see my last comment
19:52 paramat     erm
19:53 paramat     i'm not sure about concept, fairly neutral
19:54 paramat     more work when adding new flags to mapgen, and these flags are already settable in advanced settings and .conf
19:54 paramat     being selfish i would prefer not to have this
19:54 sfan5       having them in the menu would be nice
19:54 paramat     :]
19:54 sfan5       but it doesn't mean it needs to be detailed
19:55 sfan5       e.g. noiseparams should not be settable from there
19:55 paramat     agreed

srifqi committed:
Only update minetest.conf when needed

Only write setting to minetest.conf when it is different than default or the setting already there.
Update flags lists.
Change from "Close" to "Save" and "Cancel".
Don't set for nil value by assuming uses default.
Use tostring instead of dump.

srifqi commented:
Just updated with suggestion from @paramat and @SmallJoker. Will rebase when everything is okay.

ShadowNinja reviewed (changes requested):
This sets the default values in `minetest.conf` instead of setting the actual world-specific values in the world configuration file.

srifqi commented:
Maybe we can bind to create world dialog instead of working independently? I'll try to figure out how.

srifqi committed:
Only apply to minetest.conf when creating new world

Temporary flags and some function for loading and saving is now on global scope.
Change back to one button: Close.

srifqi commented:
At this point, this will be only saved to minetest.conf after player click ```Create``` and will reset back to what ```minetest.conf``` has when player click ```Cancel```. Maybe we should reset back to default instead of to what the content of minetest.conf is?

I see that we have Settings API, any suggestion for this?

paramat commented:
Currently i'm still very unsure about this, due to issues of behaviour. These settings are settable in .conf and advanced settings already so i'm not sure about a 3rd place.
It also makes working on mapgen more difficult, every new flag needs this code changing too.
So :-1: sorry.

C1ffisme commented:
@paramat This is usually where I would disagree for the sake of configuration, but to be fair, I don't think I really care about letting someone turn off biome blending, caves or trees. Some of this stuff only affects v6 now, and flat terrain is a totally different mapgen now.

Fixer-007 commented:
I would like to have possibility to configure those settings via gui during world creation.

paramat commented:
Thought on this more.
I know that the intention is good, and that some would like this, and i appreciate the amount of work that has gone into it, so i feel bad making this decision, but for several reasons i have to strongly oppose.
* World creation menu is for the most basic and essential settings only: name, seed, mapgen, and this is enough for most users.
* Mapgen global and specific flags, and the non-noise parameters, are non-essential to change and are already settable in 2 ways in .conf and advanced settings, i want to avoid a 3rd way to set these to avoid duplication and non-essential extra code.
* I feel we should instead work on advanced setting and organise the sections to be more intuitive, currently it is often guesswork, even for me, to know which initial section to expand.
* Currently mapgen files are nicely self-contained. if i add or remove a parameter i don't have to touch any files other than mapgen file, conf.example or settingtypes.txt, and those 2 text files are simple and quick to update.
With this PR i would have to completely reformat formspecs with every change, this will often be a bigger job than the mapgen work i am doing. This will be very irritating to have to do and i can tell you now i am not prepared to do that, editing just the 2 text files is already irritating and boring enough.
Some may argue that 'others will maintain the extra code and make the menu changes you don't want to', this is not practical or realistic, it is and will be down to me to make complete PRs for mapgen and i need to be able to do so.
* With our severe dev shortage we have to keep things simple if they possibly can be.

rubenwardy commented:
Any update on this?

srifqi commented:
IDK, maybe this won't be added? (From the tag set.)

paramat commented:
I am the MT mapgen dev so i know how this will affect dev, this will make my dev work much more difficult and greatly increase my workload, this is also not much needed. So i'm sorry but there is absolutely no way i can allow this to go ahead, this will be hell for me. This type of non-essential complexity is the last thing we need with our severe lack of dev time.

srifqi commented:
So, that marks the end of this more than two years old PR. Thank you so much for your attentions.

I can understand how this will make map dev complicated. Also, I agree on making better Advanced Settings.

Happy coding, all! :D

jastevenson303 commented:
I'm assuming the notion of mapgen flag _combinations_  set up as a small number of selectable _presets_ has already been brought up?

srifqi commented:
Let's move on to #3696 to make much better UI.

paramat commented:
New PR #6399

paramat commented:
I think we can possibly revive a small part of this to allow setting of a few basic global mapgen parameters during world creation. These would probably be:

`mapgen_limit = 31000`
`mg_flags = caves,dungeons,light,decorations`

Not `water_level` as changing this without re-registering a custom biome system makes a nasty mess that will give a bad impression.
Not `chunksize` as this needs understanding of the resullts before adjusting.

srifqi commented:
As #6399 already merged. Maybe we can start reviving?

paramat commented:
Yes.
